sql/core/src/main/scala/org/apache/spark/sql/execution/script/ScriptTransformationExec.scala

```diff
@@ -0,0 +1,313 @@
+    outputProjection: Projection,
+    ioschema: ScriptTransformIOSchema,
+    outputStream: OutputStream,
+    proc: Process,
+    stderrBuffer: CircularBuffer,
+    taskContext: TaskContext,
+    conf: Configuration
+  ) extends Thread("Thread-ScriptTransformation-Feed") with Logging with Serializable {
+
+  setDaemon(true)
+
+  @volatile protected var _exception: Throwable = null
+
+  /** Contains the exception thrown while writing the parent iterator to the external process. */
+  def exception: Option[Throwable] = Option(_exception)
+
+  protected def init(): Unit = {
+    TaskContext.setTaskContext(taskContext)
+  }
+
+  protected def processRow(row: InternalRow, numColumns: Int): Unit = {
+    val data = if (numColumns == 0) {
+      ioschema.inputRowFormatMap("TOK_TABLEROWFORMATLINES")
+    } else {
+      val sb = new StringBuilder
+      sb.append(row.get(0, inputSchema(0)))
+      var i = 1
+      while (i < numColumns) {
+        sb.append(ioschema.inputRowFormatMap("TOK_TABLEROWFORMATFIELD"))
+        sb.append(row.get(i, inputSchema(i)))
+        i += 1
+      }
+      sb.append(ioschema.inputRowFormatMap("TOK_TABLEROWFORMATLINES"))
+      sb.toString()
+    }
+    outputStream.write(data.getBytes(StandardCharsets.UTF_8))
+  }
+
+  override def run(): Unit = Utils.logUncaughtExceptions {
+    init()
+
+    // We can't use Utils.tryWithSafeFinally here because we also need a `catch` block, so
+    // let's use a variable to record whether the `finally` block was hit due to an exception
+    var threwException: Boolean = true
+    val numColumns = inputSchema.length
+    try {
+      iter.map(outputProjection).foreach(row => processRow(row, numColumns))
+      threwException = false
+    } catch {
+      case t: Throwable =>
+        // An error occurred while writing input, so kill the child process. According to the
+        // Javadoc this call will not throw an exception:
+        _exception = t
+        proc.destroy()
+        throw t
+    } finally {
+      try {
+        Utils.tryLogNonFatalError(outputStream.close())
+        if (proc.waitFor() != 0) {
+          logError(stderrBuffer.toString) // log the stderr circular buffer
+        }
+      } catch {
+        case NonFatal(exceptionFromFinallyBlock) =>
+          if (!threwException) {
+            throw exceptionFromFinallyBlock
+          } else {
+            log.error("Exception in finally block", exceptionFromFinallyBlock)
+          }
+      }
+    }
+  }
+}
+
+private[sql]
+object ScriptTransformIOSchema {
+  def apply(input: ScriptInputOutputSchema): ScriptTransformIOSchema = {
+    new ScriptTransformIOSchema(
+      input.inputRowFormat,
+      input.outputRowFormat,
+      input.schemaLess)
+  }
+}
+
+/**
+ * The wrapper class of Hive input and output schema properties
+ */
+private[sql] class ScriptTransformIOSchema (
+    inputRowFormat: Seq[(String, String)],
+    outputRowFormat: Seq[(String, String)],
+    schemaLess: Boolean) extends Serializable {
```

Comment:
document what these parameters mean?

e.g. what schemaLess mean, and what the seq of string tuples mean for inputRowFormat and outputRowFormat?